Mathlib/Tactic/MkIffOfInductiveProp.lean

```diff
@@ -0,0 +1,383 @@
+    let t ← inferType arg
+    let l := (← inferType t).sortLevel!
+    pure $ if arg.occurs i || l != Level.zero
+      then mkApp2 (mkConst `Exists [l] : Expr) t (←(mkLambdaFVars #[arg] i))
+      else mkApp2 (mkConst `And [] : Expr) t i)
+  inner
+
+/-- `mkOpLst op empty [x1, x2, ...]` is defined as `op x1 (op x2 ...)`.
+  Returns `empty` if the list is empty. -/
+def mkOpLst (op : Expr) (empty : Expr) : List Expr → Expr
+| []        => empty
+| [e]       => e
+| (e :: es) => mkApp2 op e $ mkOpLst op empty es
+
+/-- `mkAndLst [x1, x2, ...]` is defined as `x1 ∧ (x2 ∧ ...)`, or `True` if the list is empty. -/
+def mkAndLst : List Expr → Expr := mkOpLst (mkConst `And) (mkConst `True)
+
+/-- `mkOrLst [x1, x2, ...]` is defined as `x1 ∨ (x2 ∨ ...)`, or `False` if the list is empty. -/
+def mkOrLst : List Expr → Expr := mkOpLst (mkConst `Or) (mkConst `False)
+
+/-- Drops the final element of a list. -/
+def List.init : List α → List α
+| []     => []
+| [_]    => []
+| a::l => a::init l
+
+/-- Converts an inductive constructor `c` into an intermediate representation that
+will later be mapped into a proposition.
+-/
+def constrToProp (univs : List Level) (params : List Expr) (idxs : List Expr) (c : Name) :
+  MetaM ((List (Option Expr) × (Expr ⊕ Nat)) × Expr)  :=
+do let type := (← getConstInfo c).instantiateTypeLevelParams univs
+   let type' ← Meta.forallBoundedTelescope type (params.length) fun fvars ty => do
+     pure $ ty.replaceFVars fvars params.toArray
+
+   Meta.forallTelescope type' fun fvars ty => do
+     let idxs_inst := ty.getAppArgs.toList.drop params.length
+     let (bs, eqs, subst) := compactRelation fvars.toList (idxs.zip idxs_inst)
+     let bs' := bs.filterMap id
+     let eqs ← eqs.mapM (λ⟨idx, inst⟩ => do
+          let ty ← idx.fvarId!.getType
+          let instTy ← inferType inst
+          let u := (←inferType ty).sortLevel!
+          if ←isDefEq ty instTy
+          then pure (mkApp3 (mkConst `Eq [u]) ty idx inst)
+          else pure (mkApp4 (mkConst `HEq [u]) ty idx instTy inst))
+     let (n, r) ← match bs', eqs with
+     | [], [] => do
+           pure (Sum.inr 0, (mkConst `True))
+     | _, []  => do
+          let t : Expr ← bs'.getLast!.fvarId!.getType
+          let l := (←inferType t).sortLevel!
+          if l == Level.zero then do
+            let r ← mkExistsLst (List.init bs') t
+            pure (Sum.inl bs'.getLast!, subst r)
+          else do
+            let r ← mkExistsLst bs' (mkConst `True)
+            pure (Sum.inr 0, subst r)
+     | _, _ => do
+       let r ← mkExistsLst bs' (mkAndLst eqs)
+       pure (Sum.inr eqs.length, subst r)
+     pure ((bs, n), r)
+
+/-- Has the effect of `refine ⟨e, ?_⟩`.
+-/
+def existsi (mvar : MVarId) (e : Expr) : MetaM MVarId := do
+  let (subgoals,_) ← Elab.Term.TermElabM.run $ Elab.Tactic.run mvar do
+    Elab.Tactic.evalTactic (←`(tactic| refine ⟨?_,?_⟩))
+  let [sg1, sg2] := subgoals | throwError "expected two subgoals"
+  sg1.assign e
+  pure sg2
```

Comment:
Maybe move this to the file `Lean/Meta.lean`? And put it in the namespace `Lean.MVarId` so it can be called with dot notation?

Reply:
Done. I also made it take a list of expressions instead of a single expression, to better match the `existsi` tactic.